android/src/main/java/io/spokestack/RNSpokestack/RNSpokestackAdapter.java

```diff
@@ -0,0 +1,91 @@
+package io.spokestack.RNSpokestack;
+
+import android.os.Build;
+
+import com.facebook.react.bridge.Arguments;
+import com.facebook.react.bridge.WritableMap;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+import androidx.annotation.RequiresApi;
+import io.spokestack.spokestack.nlu.NLUResult;
+import io.spokestack.spokestack.nlu.Slot;
+import io.spokestack.spokestack.tts.TTSEvent;
+import io.spokestack.spokestack.util.EventTracer;
+
+@RequiresApi(api = Build.VERSION_CODES.N)
+public class RNSpokestackAdapter extends io.spokestack.spokestack.SpokestackAdapter {
+    public BiFunction<String, WritableMap, Void> sendEvent;
+
+    @Override
+    public void onTrace(EventTracer.Level level, String message) {
+        WritableMap react_event = Arguments.createMap();
+        react_event.putString("event", "trace");
+        react_event.putString("trace", message);
+        react_event.putString("level", level.toString());
+        sendEvent.apply("onSpeechEvent", react_event);
+    }
+
+    @Override
+    public void eventReceived(TTSEvent event) {
+        WritableMap react_event = Arguments.createMap();
+        if (event.getError() == null) {
+            react_event.putString("event", "success");
+            react_event.putString("url", event.getTtsResponse().getAudioUri().toString());
+            sendEvent.apply("onTTSEvent", react_event);
+        } else {
+            react_event.putString("event", "failure");
+            react_event.putString("error", event.getError().getLocalizedMessage());
+            sendEvent.apply("onErrorEvent", react_event);
+        }
+    }
+
+    @Override
+    public void onError(Throwable err) {
+        WritableMap react_event = Arguments.createMap();
+        react_event.putString("event", "error");
+        react_event.putString("error", err.getLocalizedMessage());
+        sendEvent.apply("onErrorEvent", react_event);
+    }
+
+    @Override
+    public void call(NLUResult arg) {
+        WritableMap reactEvent = Arguments.makeNativeMap(toEvent(arg));
+        sendEvent.apply("onNLUEvent", reactEvent);
+    }
+
+    static Map<String, Object> toEvent(NLUResult nluResult) {
```

Comment:
I totally forgot about that since it happened while I was out. I think testability is a good goal here, just one that I have spent no effort on to date.

Reply:
OK. I hate to have the extra conversion, but I haven't found a good way to test the FB maps yet.